File: leetcode/python/Convert Sorted List to Binary Search Tree/Convert_Sorted_List_to_Binary_Search_Tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    # @param head, a list node
    # @return a tree node
    def sortedListToBST(self, head):
        if not head:
            return head
        slow = head
        fast = head
        pre = None
        while fast.next and fast.next.next:
            pre = slow
            slow = slow.next
            fast = fast.next.next

        if pre is not None:
            pre.next = None
        else:
            head = None
        root = TreeNode(slow.val)
        root.left = self.sortedListToBST(head)
        root.right = self.sortedListToBST(slow.next)
        return root

from ListNode import *
from TreeNode import *
node1 = ListNode(1)
node2 = ListNode(3)
node3 = ListNode(1)

node1.next = node2

solu = Solution()
tree = solu.sortedListToBST(node1)
print 'ok'

File: leetcode/cpp/ConvertSortedListtoBinarySearchTree.cpp
//
// Created by zhangqi on 2018/1/11.
//


#include <vector>
#include "TreeNode.h"
#include "ListNode.h"
using namespace std;
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        return sortedListToBSTHelper(head, nullptr);
    }

private:
    TreeNode* sortedListToBSTHelper(ListNode* start, ListNode* end){
        if(start == end){
            return nullptr;
        }

        ListNode *slow = start;
        ListNode *fast = start;
        while(fast != end && fast->next != end){
            slow = slow->next;
            fast = fast->next->next;
        }

        TreeNode *node = new TreeNode(slow->val);
        node->left = sortedListToBSTHelper(start, slow);
        node->right = sortedListToBSTHelper(slow->next, end);
        return node;
    }
};